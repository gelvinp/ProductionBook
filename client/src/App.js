import React, { Component } from 'react'
import { Grid, Button, Header, Input, Segment } from 'semantic-ui-react'
import FileList from './FileList.js'
import APIRequest from './APIRequest.js'
import File from './File.js'
import PDFViewer from './PDFViewer.js'
import PDFJs from './pdfjs.js'

class App extends Component {
  state = {
    input: '',
    password: '',
    files: [],
    error: false,
    selected: undefined
  }
  submitPassword = () => {
    this.setState(prevState => ({ password: prevState.input }))
    APIRequest.index(this.state.input).then(json => {
      if (json.error) {
        this.setState({ error: true })
      } else {
        const files = json.data.map(file => new File(file.name, file.hash))
        console.log(files)
        this.setState({ files: files })
      }
    })
  }
  selectFile = hash => {
    console.log(`File selected: ${hash}`)
    APIRequest.request(hash, this.state.password).then(json => {
      if (json.error) {
        this.setState({ error: true })
      } else {
        console.log(json.data)
        this.setState({ selected: json.data })
      }
    })
  }
  render() {
    const { password, input, files, error, selected } = this.state
    const file =
      typeof selected !== 'undefined' ? (
        <PDFViewer backend={PDFJs} src={selected} />
      ) : (
        <p>No file selected</p>
      )
    return password.length === 0 ? (
      <div
        style={{
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center'
        }}
      >
        <Segment style={{ width: 350 }}>
          <Header>Password:</Header>
          <Input
            value={input}
            onChange={e => this.setState({ input: e.target.value })}
            type="password"
          />
          <Button onClick={this.submitPassword}>Submit</Button>
        </Segment>
      </div>
    ) : error ? (
      <div
        style={{
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center'
        }}
      >
        <Segment style={{ width: 350 }}>
          <Header>
            There was an error, check your password and try again.
          </Header>
        </Segment>
      </div>
    ) : (
      <Grid>
        <Grid.Column width={4}>
          <FileList files={files} selectFile={this.selectFile} />
        </Grid.Column>
        <Grid.Column width={12}>{file}</Grid.Column>
      </Grid>
    )
  }
}

export default App
