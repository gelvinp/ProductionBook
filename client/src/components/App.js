import React, { Fragment, Component } from 'react'
import {
  Grid,
  Responsive,
  Button,
  Header,
  Input,
  Segment,
  Form,
} from 'semantic-ui-react'
import DesktopDisplay from '../containers/DesktopDisplayContainer.js'
import MobileDisplay from '../containers/MobileDisplayContainer.js'
import PropTypes from 'prop-types'

class App extends Component {
  state = {
    id: '',
    password: '',
    error: false,
    section: '',
    sectionError: false,
  }
  getWidth = () => {
    const isSSR = typeof window === 'undefined'

    return isSSR ? Responsive.onlyTablet.minWidth : window.innerWidth
  }
  handleKeyDown = (e, func) => {
    if (e.key === 'Enter') {
      func()
    }
  }
  submitPassword = async () => {
    const { attemptLogin } = this.props
    const { id, password } = this.state
    if (id.length === 0 || password.length === 0) {
      return
    }
    const loginInfo = { id: id, password: password }
    const success = await attemptLogin(loginInfo)
    this.setState({ id: '', password: '' })
    if (!success) {
      this.setState({ error: true })
    }
  }
  componentDidMount() {
    const { attemptLogin } = this.props
    attemptLogin()
  }
  render() {
    const { id, password, error } = this.state
    const { passwordIsEmpty } = this.props
    const display = (
      <Fragment>
        <Responsive
          getWidth={this.getWidth}
          minWidth={Responsive.onlyTablet.minWidth}
        >
          <DesktopDisplay />
        </Responsive>
        <Responsive
          getWidth={this.getWidth}
          maxWidth={Responsive.onlyMobile.maxWidth}
        >
          <MobileDisplay />
        </Responsive>
      </Fragment>
    )
    return error ? (
      <div
        style={{
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <Segment style={{ width: 350 }}>
          <Header>
            There was an error, check your password and try again.
          </Header>
        </Segment>
      </div>
    ) : passwordIsEmpty ? (
      <div
        style={{
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <Segment style={{ width: 350 }}>
          <Grid>
            <Grid.Row>
              <Grid.Column>
                <Header>Login</Header>
                <Form>
                  <Form.Input
                    label="ID"
                    value={id}
                    onChange={e => this.setState({ id: e.target.value })}
                    onKeyDown={e => this.handleKeyDown(e, this.submitPassword)}
                  />
                  <Form.Input
                    label="Password"
                    value={password}
                    type="password"
                    onChange={e => this.setState({ password: e.target.value })}
                    onKeyDown={e => this.handleKeyDown(e, this.submitPassword)}
                  />
                </Form>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column>
                <Button
                  disabled={!(id.length !== 0 && password.length !== 0)}
                  onClick={this.submitPassword}
                >
                  Submit
                </Button>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        </Segment>
      </div>
    ) : (
      display
    )
  }
}

App.propTypes = {
  passwordIsEmpty: PropTypes.bool.isRequired,
  attemptLogin: PropTypes.func.isRequired,
}

export default App
