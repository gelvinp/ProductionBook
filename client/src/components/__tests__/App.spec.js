import React from 'react'
import App from '../App.js'
import { shallow } from 'enzyme'

describe('App', () => {
  it('Displays the password dialogue when empty', () => {
    const wrapper = shallow(
      <App
        passwordIsEmpty={true}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={jest.fn()}
      />
    )
    expect(wrapper.find('FormInput').length).toBe(2)
  })

  it('Updates the password field', () => {
    const wrapper = shallow(
      <App
        passwordIsEmpty={true}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={jest.fn()}
      />
    )
    wrapper
      .find('FormInput')
      .at(0)
      .simulate('change', { target: { value: 'Test' } })
    wrapper
      .find('FormInput')
      .at(1)
      .simulate('change', { target: { value: 'TestTwo' } })
    expect(wrapper.state('id')).toEqual('Test')
    expect(wrapper.state('password')).toEqual('TestTwo')
  })

  it('Submits password on button click', () => {
    const attemptLogin = jest.fn().mockImplementation(() => {
      return Promise.resolve(false)
    })
    const wrapper = shallow(
      <App
        passwordIsEmpty={true}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={attemptLogin}
      />
    )
    wrapper.setState({ id: 'a', password: 'a' })
    wrapper.find('Button').simulate('click')
    process.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.state('error')).toBeTruthy()
    })
  })

  it('Submits password on Enter key down', () => {
    const attemptLogin = jest.fn().mockImplementation(() => {
      return Promise.resolve(false)
    })
    const wrapper = shallow(
      <App
        passwordIsEmpty={true}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={attemptLogin}
      />
    )
    wrapper.setState({ id: 'a', password: 'a' })
    wrapper
      .find('FormInput')
      .at(1)
      .simulate('keydown', { key: 'Enter' })
    process.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.state('error')).toBeTruthy()
    })
  })

  it('Doesnt submit password on non Enter key down', () => {
    const attemptLogin = jest.fn().mockImplementation(() => {
      return Promise.resolve(false)
    })
    const wrapper = shallow(
      <App
        passwordIsEmpty={true}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={attemptLogin}
      />
    )
    wrapper
      .find('FormInput')
      .at(0)
      .simulate('keydown', { key: 'Down' })
    process.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.state('error')).not.toBeTruthy()
    })
  })

  it('Doesnt submit password when id or password are empty', async () => {
    const attemptLogin = jest.fn().mockImplementation(() => {
      return Promise.resolve(false)
    })
    const wrapper = shallow(
      <App
        passwordIsEmpty={true}
        setPassword={jest.fn()}
        attemptLogin={attemptLogin}
      />
    )
    wrapper.setState({ id: 'a' })
    await wrapper.instance().submitPassword()
    wrapper.setState({ id: '', password: 'a' })
    await wrapper.instance().submitPassword()
    process.nextTick(() => {
      expect.assertions(1)
      expect(wrapper.state('error')).not.toBeTruthy()
    })
  })

  it('Accepts a valid password', async () => {
    let match = undefined
    const attemptLogin = jest.fn().mockImplementation(pass => {
      expect(pass).toEqual(match)
      return Promise.resolve(true)
    })
    const wrapper = shallow(
      <App passwordIsEmpty={true} attemptLogin={attemptLogin} />
    )
    wrapper.setState({ id: 'Test', password: 'Password' })
    match = { id: 'Test', password: 'Password' } // Put this down here because attemptLogin gets called on mount
    await wrapper.instance().submitPassword()
    expect(wrapper.state('error')).not.toBeTruthy()
    expect.assertions(3)
  })

  it('Shows an error page', () => {
    const wrapper = shallow(
      <App
        passwordIsEmpty={false}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={jest.fn()}
      />
    )
    wrapper.setState({ error: true })
    expect(
      wrapper
        .find('Header')
        .children(0)
        .text()
    ).toEqual('There was an error, check your password and try again.')
  })

  it('Shows the display', () => {
    const wrapper = shallow(
      <App
        passwordIsEmpty={false}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={jest.fn()}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('Gets the width', () => {
    const wrapper = shallow(
      <App
        passwordIsEmpty={false}
        setPassword={jest.fn()}
        createSection={jest.fn()}
        createDocument={jest.fn()}
        attemptLogin={jest.fn()}
      />
    )
    expect(wrapper.instance().getWidth()).toBe(1024)
  })
})
