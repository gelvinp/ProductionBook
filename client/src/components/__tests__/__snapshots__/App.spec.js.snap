// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Shows the display 1`] = `
<Fragment>
  <Responsive
    getWidth={[Function]}
    minWidth={768}
  >
    <Connect(DesktopDisplay) />
  </Responsive>
  <Responsive
    getWidth={[Function]}
    maxWidth={767}
  >
    <Connect(MobileDisplay) />
  </Responsive>
</Fragment>
`;
